### PR TITLE
Macroquad loop

### PR DESCRIPTION
**Pull request by vicentevieytes**
Implements the main game loop on main.rs.

Extends the Game API with the following functions:
```rust
    pub fn from_size_and_cells(dimensions: Position, cells: &[Position],
    ) -> Result<Self, GameError> 
    pub fn of_size(dimensions: Position) -> Self 
    pub fn toggle_cell(&mut self, pos: (usize, usize)) -> Result<(), GameError>
    pub fn kill(&mut self, pos: Position) -> Result<(), GameError> 
    pub fn give_life(&mut self, pos: Position) -> Result<(), GameError> 
    pub fn kill_list(&mut self, list: &[Position]) -> Result<(), GameError> 
    pub fn give_life_list(&mut self, list: &[Position]) -> Result<(), GameError> 
    pub fn alive_cells(&self) -> &HashSet<Position> 
    pub fn dimensions(&self) -> Position 
    pub fn is_alive(&self, pos: Position) -> bool 
```

Also includes several reformats, including variable and functino renames and method abstractions, let me know if anything is hard to understand.

### CODE REVIEW

File: src/game.rs [hunk 1]
```diff
@@ -41,17 +41,28 @@ impl Game {
...
     }
 
-    pub fn toggle_cell(&mut self, pos: (usize, usize)) {
+    /// Kills the cell if it's alive, gives life to it if it's dead.
+    pub fn toggle_cell(&mut self, pos: (usize, usize)) -> Result<(), GameError> {
+        if !Self::in_bounds(pos.0, pos.1, self.grid.len(), self.grid[0].len()) {
+            return Err(GameError::OutOfBoundsGridAccess(
+                (pos.0, pos.1),
+                (self.grid.len(), self.grid[0].len()),
+            ));
```
Comment (vicentevieytes) [inline src/game.rs:50]: This should be self.in_bounds(position), i will be refactoring it

File: src/game.rs [hunk 2]
```diff
@@ -22,22 +22,60 @@ impl Game {
...
+                ));
             }
             //Otherwise, give life to the cell and return OK
-            grid[*cell_i][*cell_j].give_life();
+            grid[*cell_row][*cell_column].give_life();
         }
         let alive_cells = cells.clone();
 
         Ok(Game { grid, alive_cells })
     }
```
Comment (vicentevieytes) [inline src/game.rs]: This variable name formatting was done to fix a bug, where i was not using the indexes corretly. It should be on a different PR but I couldn't figure out how to do that.

File: src/game.rs [hunk 3]
```diff
@@ -22,22 +22,66 @@ impl Game {
...
+        let (i, j) = pos;
+
+        let cell = &mut self.grid[i][j];
+
+        // Kills the cell, if it's on self.alive_cells, remove it.
+        if cell.is_alive() {
+            cell.kill();
+            if let Some(index) = self.alive_cells.iter().position(|value| *value == pos) {
+                self.alive_cells.swap_remove(index);
+            }
```
Comment (vicentevieytes) [inline src/game.rs]: This checks if the position is already on alive_cells and removes it, It should be abstracted into a function.

File: src/game.rs [hunk 4]
```diff
@@ -22,22 +22,66 @@ impl Game {
...
 
+    /// Kills the cell if it's alive, gives life to it if it's dead.
+    pub fn toggle_cell(&mut self, pos: (usize, usize)) -> Result<(), GameError> {
+        if !Self::in_bounds(pos.0, pos.1, self.grid.len(), self.grid[0].len()) {
+            return Err(GameError::OutOfBoundsGridAccess(
+                (pos.0, pos.1),
+                (self.grid.len(), self.grid[0].len()),
+            ));
+        }
+
```
Comment (vicentevieytes) [inline src/game.rs]: This out of bounds error check should also be abstracted into a function

File: src/game.rs (hunk 2 above)
Comment (vicentevieytes) [reply src/game.rs]: In_bounds check at line and error return at line 30 should be abstracted into a function.

File: src/game.rs [hunk 5]
```diff
@@ -22,22 +22,66 @@ impl Game {
...
+    //// Returns the vector of alive cells at the current iteration
+    pub fn alive_cells(&self) -> &Vec<Position> {
+        &self.alive_cells
+    }
+
+    /// Returns the width and height of the grid.
+    pub fn dimensions(&self) -> Position {
+        (self.grid.len(), self.grid[0].len())
+    }
+
```
Comment (vicentevieytes) [inline src/game.rs:94]: This function is not being used. It should be used to check for in_bounds and for grid access in general.

File: src/game.rs [hunk 6]
```diff
@@ -103,26 +147,26 @@ impl Game {
...
 
-    fn in_bounds<T>(i: T, j: T, w: T, h: T) -> bool
+    fn in_bounds<T>(row: T, column: T, height: T, width: T) -> bool
     where
         T: Num + PartialOrd,
     {
-        i >= T::zero() && j >= T::zero() && i < h && j < w
+        row >= T::zero() && column >= T::zero() && row < height && column < width
     }
 }
```
Comment (vicentevieytes) [inline src/game.rs:187]: More variable name reformating, the names before where not declarative.

File: src/game.rs [hunk 7]
```diff
@@ -2,40 +2,99 @@ use crate::cell::Cell;
...
+    pub fn kill(&mut self, pos: Position) -> Result<(), GameError> {
+        self.check_in_bounds(pos)?;
+        let cell = &mut self.grid[pos.0][pos.1];
+        cell.kill();
+        self.alive_cells.remove(&pos);
+        Ok(())
+    }
+
+    /// Gives life to a specific cell, returns error when out of bounds
+    pub fn give_life(&mut self, pos: Position) -> Result<(), GameError> {
```
Comment (MarcosNicolau) [inline src/game.rs:55]: why do you have a Result here?

Comment (MarcosNicolau) [reply src/game.rs:55]: Oh alright for check_in_bounds

File: src/game.rs [hunk 8]
```diff
@@ -2,40 +2,99 @@ use crate::cell::Cell;
...
+    /// Gives life to a list of cells
+    pub fn give_life_list(&mut self, list: &[Position]) -> Result<(), GameError> {
+        for &pos in list.iter() {
+            self.give_life(pos)?;
         }
-        let alive_cells = cells.clone();
+        Ok(())
+    }
+
+    pub fn genocide(&mut self) {
```
Comment (MarcosNicolau) [inline src/game.rs:79]: 🤣

File: src/game.rs (hunk 7 above)
Comment (MarcosNicolau) [reply src/game.rs:55]: do you really need to do the bound checking? I mean, if the you are out of bounds you have a bug in your program, but if everything is correct, then why would you want to check that?

**Review by MarcosNicolau — COMMENTED** [review]
Other than that, it looks pretty alright to me

File: src/game.rs (hunk 7 above)
Comment (vicentevieytes) [reply src/game.rs:55]: It really helped me find bugs when i was developing the game loop.

This way you don't allow any of the methods of Game to panic, you leave the responsibility of handling the errors to the caller, in this case the loop in main.rs. If anybody else wanted to implement a game which used this same module, the errors would indicate to them if their implementation has indexing errors.

It's a little bit annoying to read, but I thought it was better this way, i might be wrong tho. I also just wanted to practice error definitions and handling
